pkg/blob/controllerserver.go

```diff
@@ -779,6 +788,53 @@ func (d *Driver) copyVolume(ctx context.Context, req *csi.CreateVolumeRequest, a
 	}
 }
 
+func (d *Driver) authorizeAzcopyBySecurityPrincipal() error {
+	if len(d.cloud.Config.AADClientSecret) > 0 {
+		klog.V(2).Infof("use service principal to authorize azcopy")
+		os.Setenv("AZCOPY_AUTO_LOGIN_TYPE", "SPN")
+		if err := os.Setenv("AZCOPY_AUTO_LOGIN_TYPE", "SPN"); err != nil {
+			klog.Errorf("failed to set AZCOPY_AUTO_LOGIN_TYPE=SPN, error: %v", err)
+			return err
+		}
+		if d.cloud.Config.AADClientID == "" || d.cloud.Config.AADClientSecret == "" || d.cloud.Config.TenantID == "" {
+			return fmt.Errorf("AADClientID: %s, AADClientSecret: %s and TenantID: %s must be set when use federated workload identity extension", d.cloud.Config.AADClientID, d.cloud.Config.AADClientSecret, d.cloud.Config.TenantID)
+		}
+		if err := os.Setenv("AZCOPY_SPA_APPLICATION_ID", d.cloud.Config.AADClientID); err != nil {
+			klog.Errorf("failed to set AZCOPY_SPA_APPLICATION_ID, error: %v", err)
```

Comment:
no need to log every error msg, `return err` is enough

Reply:
done